proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+  as an identity provider. Offer configuration options for extracting usernames and group memberships, as well
+  as for preventing authentication unless a user belongs to a certain GitHub organization.
+- Allow GitHubIdentityProvider(s) to be configured by admins on FederationDomains, and enable all features of
+  FederationDomains on them (e.g. identity transformations).
+- Allow end users to use a web browser-based authentication flow, similar to other supported identity providers,
+  where the end user only needs to authenticate using a web browser once a day.
+- Allow end users to use a CLI-based authentication flow which does not require a web browser,
+  similar to other supported identity providers. This enables using Pinniped authentication for scripting
+  and CI purposes. Allow the admin to disable this feature if they prefer.
+- Regardless of which flow the user uses to authenticate, regularly check with GitHub to see if the user should
+  be allowed to continue their session. If the user's GitHub tokens were revoked, or if the user no longer belongs
+  to the configured GitHub organization(s), then end their Pinniped Supervisor session to block their continued
+  access to Kubernetes clusters.
+- Regardless of which flow the user uses to authenticate, regularly check with GitHub to see if the user's group
+  memberships have changed and update them accordingly.
+- Any client of the Supervisor should be able to authenticate GitHub users using the web browser-based
+  authentication flow, including the Pinniped CLI and third-party clients registered with the Supervisor as OIDCClients.
+- The IDP chooser page and the IDP discovery endpoints should both include the configured GitHub identity provider(s)
+  for the FederationDomain in the lists that they return.
+- Avoid requiring any changes to the Pinniped CLI or other clients. From a client's point of view,
+  which external identity provider is used for authentication should be transparent.
+
+### Specification / How it Solves the Use Cases
+
+#### New GitHubIdentityProvider resource
+
+Add a new CRD called GitHubIdentityProvider. Here is an example:
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: GitHubIdentityProvider
+metadata:
+  namespace: supervisor
+  name: github
+spec:
+  github_api:
+    # Required only for GitHub Enterprise Server (on-prem).
+    # Defaults to using the GitHub's public API.
+    host: my.example.com
+    # X.509 Certificate Authority (base64-encoded PEM bundle).
+    # If omitted, a default set of system roots will be trusted.
+    # Required only for GitHub Enterprise Server (on-prem), and only
+    # when its CA will not be trusted by the default system roots.
+    certificateAuthorityData: LS0tLS1CRUdJTiBDRV...0tLS0tCg==
+  claims:
+    # Which property of the GitHub user record shall determine
+    # the username in Kubernetes. Can be either "id", "login",
```

Comment:
Could they use a CEL transformation to further manipulate the username?

Reply:
Yes, as mentioned a few lines down in the same comment:

```
     # If desired, an admin could configure identity
     # transformation expressions on a FederationDomain to
     # further customize these usernames.
```